### DISCUSSION TITLE
WebAuthN prioritized over username/password form - is it possible?

### DISCUSSION THREAD

**Discussion by oculos** [keycloak/keycloak#24883] [Q&A]
Hi,

I have an authentication workflow so that I can use both webauthn or passwords to authenticate to keycloak.

<img width="1175" alt="Screenshot 2023-11-21 at 14 22 52" src="https://github.com/keycloak/keycloak/assets/6791923/8e65ef84-1da0-41c8-8af8-1a84115746da">

However, even if WebAuthn before username/password form on my authentication flow, I still get username/password form before.

Is this a bug, or is there a way to make Webauthn offered first and make username/password available when choosing "try other ways" instead of the opposite?

**Comment by darius-m** [comment]
I think I remember an older discussion mentioned that the order the authentication forms are presented to the user depends on the order of the credentials in the user's account. As such, if the password credential is the first, the username / password form will be presented first (disclaimer - I'm not quite sure if it also applies to the password credential, but it's unlikely it is handled differently by the selection process). You can test if the behavior changes by changing the order of the credentials in the administration page - there is even a dedicated [REST API endpoint](https://www.keycloak.org/docs-api/23.0.0/rest-api/index.html#_post_adminrealmsrealmusersidcredentialscredentialidmovetofirst) for this purpose.

**Reply by oculos** [reply]
Thanks @darius-m 
I am not quite sure how you do that, at least not at the web gui. 
I don't see the password as a movable element: 

<img width="1508" alt="Screenshot 2023-11-24 at 13 50 43" src="https://github.com/keycloak/keycloak/assets/6791923/272b450f-0caf-4287-98cb-57f71f994ebc">

**Reply by gitwittidbit** [reply]
> I think I remember an older discussion mentioned that the order the authentication forms are presented to the user depends on the order of the credentials in the user's account. As such, if the password credential is the first, the username / password form will be presented first (disclaimer - I'm not quite sure if it also applies to the password credential, but it's unlikely it is handled differently by the selection process). 

sorry, I don't understand how this would work: Prior to authentication - who is the user who's account defines the flow? Keycloak doesn't know who is going to authenticate, does it?

But I, too, would like this riddle solved. Has anyone found a solution?

**Comment by embesozzi** [comment]
Hi @oculos,
The best user experience is achieved with [WebAuth Conditional UI](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI) or Passkey autofill. The default “modal” experience might not be good in some cases, but utilizing the passkey with autofill (WebAuth Conditional UI) feature can significantly enhance the login process.
For more information and hands-on experience, check out the workshop that demonstrates the Passkey autofill experience and various MFA scenarios:
- https://github.com/embesozzi/keycloak-workshop-stepup-mfa-biometrics

**Comment by oculos** [comment]
Thanks Martin!

But that’s not in production yet, right?

**Reply by embesozzi** [reply]
In relation to Keycloak, in my case, it's still in QA. However, I implemented a similar functionality in ForgeRock, and so far, I haven't encountered many problems.
